final merge all parts

yahelroro0196 opened the pull request.

yahelroro0196 committed:
initial project creation

yahelroro0196 committed:
initial design for project.

yahelroro0196 committed:
added main java file

Commit:
updated Design graph

Commit:
updated design graph

Commit:
1: built config loader and parser
2: built menu call from main
3: built first menu loop logic
4: built input builder and validator

Commit:
added valid operators config file

Commit:
1: changed from navigable map to ArrayList<Pair<>> to support repeating numbers.
2: added calculation logic from infix to postfix

Commit:
1: changed from navigable map to ArrayList<Pair<>> to support repeating numbers.
2: added calculation logic from infix to postfix

yahelroro0196 committed:
1: Implemented complete shunting yard algorithm infix->postfix equation
2: Implemented postfix->result
3: Clean code with constants and short functions
4: Implemented abstract operator.java class + all the different operators
5: created OperatorFactory.
6: changed Pair attributes to better reflect purpose
7: Constants class

yahelroro0196 committed:
1: separated infix to postfix logic from postfix to result logic to same package but different classes.
2: more constant separation
3: added exit function by entering an empty input twice.

yahelroro0196 committed:
1: added more functions to separate and make the code cleaned and more understandable
2: added an InputFlow package to separate input parsing to it's own area
3: move config loading to CalculatorGui to clean menu.

yahelroro0196 committed:
deleted useless space in Main.java

Commit:
removed unused imports in main.

Commit:
updated README.md file.

yahelroro0196 committed:
Delete ValidOperators.properties

yahelroro0196 committed:
new line at end of file

yahelroro0196 committed:
new line at end of file

yahelroro0196 committed:
new line at end of file

Commit:
1: updated package name
2: separated GUI and Logic further
3: Added Logging and Log4j component
4: Changed equation array to <String, Type> with ENUM support
5: added 'Type' enum
6: added operator delimiter constant
7: separated constants from enums
8: cleaned Operator
9: changed operator factory to have default and unreachable case
10: added maven
11: added detection of incorrect equation
12: added log4j config file
13: added log4j log file (log.out)

Commit:
deleted useless lines in pom.xml

Commit:
Merge remote-tracking branch 'origin/calculator_part_1' into calculator_part_1

yahelroro0196 committed:
cleaned properties and gitignore

yahelroro0196 committed:
cleaned properties and gitignore

yahelroro0196 committed:
added division by zero handling.

yahelroro0196 committed:
added support to brackets and fixed InfixToPostfix parsing.

yahelroro0196 committed:
cleaned directory

Commit:
updated config to match part 1.

Commit:
1: changed Pair to javafx.util.Pair
2: got rid of  Constants.java
3: changed ConfigLoader to StreamReader

Commit:
added equation parsing and calculation unit testing!

Commit:
spacing

Commit:
added example of edge case test in separate class.

Commit:
deleted unused variable

Commit:
added cleaner exception handling to zero division, empty brackets  and more.

Commit:
added exception handling of broken bracket form.

Commit:
added switch case 'break;'

yahelroro0196 committed:
1: added cleaner unit testing code
2: divided tests between classes
3: added easier input, a new parser to parse a string equation.
4: better variable names
5: created a class for each edge case

yahelroro0196 committed:
switched to expression lambda.

yahelroro0196 committed:
1: added more unit testing cases
2: cleaned useless classes and instead added JUnit annotations to the tests
3: fixed test parser to work with multiple digit numbers
4: changed string equation input to be with spaces

yahelroro0196 committed:
1. got rid of overlaying useless package
2. added more edge cases
3. changed variable namings to be clearer with UPPERCASE

yahelroro0196 committed:
1. added config class to hold constants and paths.
2. changed some namings to camelCase
3. updated the unit tests with those variables from the config
class.

yahelroro0196 committed:
Added a more generic parsing mechanism.

yahelroro0196 committed:
Merge Part 1 to Dev

Calculator Part 1

yahelroro0196 committed:
1: cleaned up a lot of logic caused by entering input line after line
2: changed input to be one liner equation with spaces between tokens
3: generic equation parsing
4: generic constants usage

yahelroro0196 committed:
mega update:
1: moved config loading to config class so access is for everyone to use
2: updated shunting yard to allow functions assigned to a symbol like cosine/sine
3: updated unit testing to accommodate that algorithm update
4: update display menu to showcase possible equation and how to write them
5: added function factory, type
6: updated equation parsing to include either with or without space bars

yahelroro0196 committed:
Merge pull request #2 from yahelroro0196/calculator_part_2

Calculator Part 2

yahelroro0196 committed:
fixed some outdated unit tests

yahelroro0196 committed:
deleted useless overlaying package.

yahelroro0196 committed:
1: added an option to type a name of a .json file in a specific format with equations and solve all of them and save in the file.
2: cleaned folder hierarchy after adding new enums
3: added a menu with options to choose between manual equation, through a file or exit.

yahelroro0196 committed:
1: simplified json format
2: corrected invalid input case in menu

yahelroro0196 committed:
added unit test for the equation logic

yahelroro0196 committed:
cleaning imports

yahelroro0196 committed:
Merge pull request #3 from yahelroro0196/calculator_part_3

Calculator Part 3

ik8 commented on log.out:
```diff
@@ -0,0 +1,23 @@
+2020-04-21 13:26:12 WARN  Log4j:14 - Invalid input! reenter the input
```
remove the file

ik8 commented on src/main/java/CalculatorTesting/TestTypes/EquationFileCases/BadFormat.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Test
+    public void test() {
+        assertThrows(BAD_FORMAT_OUTPUT, () -> loadEquationFile(BAD_FORMAT_INPUT));
```
what is this

ik8 commented on src/main/java/CalculatorTesting/TestTypes/InfixToPostfixCases/DoubleBrackets.java:
```diff
@@ -0,0 +1,46 @@
+    private static ArrayList<Pair<String, Type>> DOUBLE_BRACKETS_OUTPUT;
+
+    @Test
+    public void test() {
```
why do you have every test with test & setup function? why not use a base class?

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/CalculatorGui.java:
```diff
@@ -0,0 +1,154 @@
+        displayOptions();
+    }
+
+    public static void displayValidOperators() {
```
where is your logger/

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,64 @@
+        JSONParser parser = new JSONParser();
+        FileWriter fileWriter = null;
+        try {
+            Object fileReader = parser.parse(new FileReader(RELATIVE_FILE_PATHS + fileName + JSON_END));
```
StreamReader!!!!!!!!!!!!!!

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,64 @@
+            iterateEquation(jsonOfEquation);
+            fileWriter.write(jsonOfEquation.toJSONString());
+        } catch (FileNotFoundException exception) {
+            throw new FileNotFoundException(Type.FILE_NOT_FOUND.name());
```
why do you call the enum type?

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/CalculatorLogic.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private static JSONObject iterateEquation(JSONObject jsonOfEquations) {
+        for (Object currEquationKey : jsonOfEquations.keySet()) {
```
you should load text files not json as I recall

ik8 commented on src/main/java/CalculatorTesting/TestTypes/PostfixToResultCases/DoubleOperator.java:
```diff
@@ -0,0 +1,46 @@
+import static CalculatorTesting.ParsePostfixString.parse;
+import static org.junit.Assert.assertThrows;
+
+public class DoubleOperator {
```
It is okay  that you type one test in class but usually there are many tests for one, and the test name should indicate what the test does, it depends what language you are using, In Java you have a suffix of 'Tests' to you class name:
```
public class OpertorTests {
  @Test
  public void doubleOpertorThrowsInputMismatch {
    // Arrange
   code = code;

    // Act
    callMethod();

    // Assert
    this.assertEquals(actValue, expectedValue); // or in different order
}
```

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/Functions/FunctionFactory.java:
```diff
@@ -0,0 +1,14 @@
+    public Function factory(String function) {
+        switch (function) {
+            case "$":
+                return new Sines();
```
just curios, why not sin and cos?

ik8 commented on src/main/java/ShuntingYardCalculator/ExceptionType.java:
```diff
@@ -0,0 +1,11 @@
+    public static final String INVALID_EQUATION_ERROR = "INVALID_EQUATION_ERROR";
+    public static final String INVALID_BRACKETS_ERROR = "INVALID_BRACKETS_ERROR";
+
+    private ExceptionType() {
```
why do you need the constructor?

ik8 commented on src/main/java/ShuntingYardCalculator/Logging/Log4j.java:
```diff
@@ -0,0 +1,44 @@
+
+import org.apache.log4j.Logger;
+
+public class Log4j {
```
one or two functions only, no need for function for each message

ik8 commented on src/main/resources/BadFormatEquations.json:
```diff
@@ -0,0 +1,6 @@
+{
```
dont use json

yahelroro0196 committed:
removed log.out file

yahelroro0196 committed:
1: merged unit tests to be like arrange act assert pattern
2: changed equation file loading to stream
3: changed equation file to .txt

yahelroro0196 committed:
Merge pull request #5 from yahelroro0196/calculator_part_3

Calculator part 3

yahelroro0196 committed:
added support to named equations with functions f.e:
sin(5)*5*(10+2)

yahelroro0196 committed:
Merge pull request #6 from yahelroro0196/calculator_part_3

added support to equations with named functions f.e sin,cos

yahelroro0196 committed:
1: fixed infix to postfix converter
2: fixed postfix evaluator
3: fixed functions parsing
4: added symbol class to prevent duplicate code in some places
5: fixed isHigherPrecedence check in logic
6: updated parsing to work better on trigo functions, brackets

yahelroro0196 committed:
Merge pull request #7 from yahelroro0196/calculator_part_3

Calculator part 3 with working sin cos tan functions

yahelroro0196 committed:
cleaned code + update menu text

Commit:
changed parsing of functions to be more generic

ik8 commented on src/main/java/ShuntingYardCalculator/Calculator/EquationParser.java:
```diff
@@ -0,0 +1,65 @@
+        return equation;
+    }
+
+    private static String iterateEquation(ArrayList<String> splitEquation, ArrayList<Pair<String, Type>> equation) {
```
monster function, split it

ik8 commented on src/main/java/ShuntingYardCalculator/OptionsFactory.java:
```diff
@@ -0,0 +1,20 @@
+public class OptionsFactory {
+    public static MenuOption factory(String chosenOption) {
+        switch (chosenOption) {
+            case "1":
```
you can pass a string to your choice to the enum, It does not have to be in range from 1 to infinity

ik8 commented on src/main/java/CalculatorTesting/InfixToPostfixTests.java:
```diff
@@ -0,0 +1,46 @@
+
+        try {
+            InfixToPostfix.infixToPostfix(doubleEmptyInput);
+        } catch (ArithmeticException e) {
```
https://stackoverflow.com/questions/156503/how-do-you-assert-that-a-certain-exception-is-thrown-in-junit-4-tests

Commit:
1: split up EquationParser function to multiple smaller ones
2: updated exception throwing tests
3: updated options in menu to be words and not  numbers.
